httpcore/_async/http_proxy.py

```diff
@@ -140,47 +133,52 @@ def __init__(
         connection = await self._get_connection_from_pool(origin)
 
         if connection is None:
-            connection = AsyncHTTPConnection(
-                origin=origin, http2=False, ssl_context=self._ssl_context,
+            # First, create a connection to the proxy server
+            proxy_connection = AsyncHTTPConnection(
+                origin=self.proxy_origin, http2=False, ssl_context=self._ssl_context,
             )
             async with self._thread_lock:
                 self._connections.setdefault(origin, set())
-                self._connections[origin].add(connection)
+                self._connections[origin].add(proxy_connection)
 
-            # Establish the connection by issuing a CONNECT request...
+            # Issue a CONNECT request...
 
             # CONNECT www.example.org:80 HTTP/1.1
             # [proxy-headers]
             target = b"%b:%d" % (url[1], url[2])
             connect_url = self.proxy_origin + (target,)
-            connect_headers = self.proxy_headers
-            proxy_response = await connection.request(
-                b"CONNECT", connect_url, headers=connect_headers, timeout=timeout
+            headers = self.proxy_headers + ([] if headers is None else headers)
+            proxy_response = await proxy_connection.request(
+                b"CONNECT", connect_url, headers=headers, timeout=timeout
             )
             proxy_status_code = proxy_response[1]
             proxy_reason_phrase = proxy_response[2]
             proxy_stream = proxy_response[4]
 
-            # Ingest any request body.
-            await read_body(proxy_stream)
+            # Read the response data without closing the socket
+            _ = b"".join([chunk async for chunk in proxy_stream])
 
             # If the proxy responds with an error, then drop the connection
             # from the pool, and raise an exception.
             if proxy_status_code < 200 or proxy_status_code > 299:
-                async with self._thread_lock:
-                    self._connections[connection.origin].remove(connection)
-                    if not self._connections[connection.origin]:
-                        del self._connections[connection.origin]
                 msg = "%d %s" % (proxy_status_code, proxy_reason_phrase.decode("ascii"))
                 raise ProxyError(msg)
 
-            # Upgrade to TLS.
-            await connection.start_tls(target, timeout)
+            # Create a new connection to the target
```

Comment:
I would add a comment here about how a successful CONNECT call results in SWITCHED PROTOCOLS, which is what justifies the creation of a new connection instead of reusing the proxy connection.

Suggestion:

```suggestion
            # The CONNECT request is successful, so we have now SWITCHED PROTOCOLS.
            # This means the proxy connection is now unusable, and we must create
            # a new one for regular requests, making sure to use the same socket to
            # retain the tunnel.
```